{% extends "parent_display.html" %}
{% block title %}Updates{% endblock %}
{% block content %}
  <h1 class="centered">Updates</h1>
  <br/>
  <h2>Version 1.6: Adds 23 new models & more!</h2>
  <p>
    <i>Sep. 15, 2019</i>
  </p>
  <p>
    Version 1.6 introduces a significant number of new models to the repository.
    We describe the addition of 23 new models that valuably expand the portion
    of the phylogenetic tree covered by BiGG Models. The models include updates
    for four <i>E. coli</i> strains, <i>S. aureus</i>, and <i>M. tuberculosis</i> H37Rv
    and new models for five Plasmodium strains, stage specific models for <i>Trypanosoma
    cruzi</i>, <i>C. difficile</i>, and <i>A. baumannii</i>. We also expand the functionality for
    hosting multi-strain models, which have proven to be insightful in a variety
    of studies centered on comparisons of related strains. These multistrain
    collections are available for download on the model pages for
    <a href="/models/iYL1228">iYL1228</a>, <a href="/models/STM_v1_0">STM_v1_0</a>,
    and <a href="/models/iCN718">iCN718</a>.
  </p>
  <a href="https://github.com/SBRG/bigg_models/releases/tag/1.6.0">Github Release</a>
  <br/>
  <br/>
  <h2>Version 1.5: Introducing Recon3D</h2>
  <p>
    <i>Feb. 26, 2018</i>
  </p>
  <p>
    Version 1.5 introduces <a href="/models/Recon3D">Recon3D</a>,
    the latest genome-scale model of the human metabolic network.
    <a href="http://www.ncbi.nlm.nih.gov/pubmed/29457794">Brunk et al.</a> describe
    how Recon3D incorporates three-dimensional metabolite and protein structure
    data for integrated analysis of human metabolism. Recon3D includes over
    5,000 metabolites, 10,000 reactions, and 2,000 human genes. Additional
    model details and features are available at <a href="http://vmh.life">http://vmh.life</a>.
  </p>
  <a href="https://github.com/SBRG/bigg_models/releases/tag/v1.5.0">Github Release</a>
  <br/>
  <br/>
  <h2>Version 1.4: Four new models, including the latest for <i>E.&nbsp;coli</i> K-12 MG1655</h2>
  <p>
    <i>Oct. 10, 2017</i>
  </p>
  <p>
    Version 1.4 introduces four models to BiGG, including the latest
    genome-scale model of <i>Escherichia coli</i> K-12 MG1655 from the Systems
    Biology Research Group at UCSD. In addition, a new universal model can be
    downloaded in JSON format from the <a href="/data_access">Data Access</a>
    page, and a number of <a
    href="https://github.com/SBRG/bigg_models/releases/tag/v1.4.0">bug fixes and
    improvements</a> are included.
  </p>
  <p>
    To account for newly discovered metabolic capabilities in
    <i>E.&nbsp;coli</i> K-12 MG1655,
    <a href="https://doi.org/10.1038/nbt.3956">Monk and Lloyd et al.</a>
    have developed an updated genome-scale model,
    <a href="/models/iML1515"><i>i</i>ML1515</a>,
    that bridges systems and
    structural biology. The new model was extensively validated using omics
    datasets and boasts a 3.7% improvement in gene essentiality predictions over
    the previous model, <i>i</i>JO1366, along with 184 new genes and 196 new
    reactions.  Check out the
    <a href="https://doi.org/10.1038/nbt.3956">manuscript</a> [<a href="http://rdcu.be/wFqN">PDF</a>]
    to see how this model can be applied to
    analyze the metabolic capabilities <i>E.&nbsp;coli</i> strains in
    metagenomic samples or used to build draft reconstructions of freshly
    sequenced clinical isolates.
  </p>
  <p>
    Other new models are: <a href="/models/iLB1027_lipid"><i>i</i>LB1027_lipid</a>
    for <i>Phaeodactylum tricornutum</i> described by <a
    href="http://www.ncbi.nlm.nih.gov/pubmed/27152931">Levering et al.</a>,
    <a href="/models/iJB785"><i>i</i>JB785</a> for <i>Synechococcus elongatus</i> described by
    <a href="http://www.ncbi.nlm.nih.gov/pubmed/27911809">Broddrick et al.</a>,
    and <a href="/models/iNF517"><i>i</i>NF517</a> for <i>Lactococcus lactis</i> described by
    <a href="http://www.ncbi.nlm.nih.gov/pubmed/23974365">Flahaut et al.</a>.
  </p>
  <a href="https://github.com/SBRG/bigg_models/releases/tag/v1.4.0">Github Release</a>
  <br/>
  <br/>
  <h2>Version 1.3: Introducing the <i>i</i>CHO model</h2>
  <p>
    <i>Nov. 23, 2016</i>
  </p>
  <p>
    Version 1.3 introduces a new model and includes a number of bug fixes.
  </p>
  <p>
    To understand the pathways underlying cell growth and protein production in
    Chinese hamster ovary (CHO) cells, the production cell line for most
    recombinant protein drugs,
    <a href="http://www.cell.com/cell-systems/fulltext/S2405-4712(16)30363-5">Hefzi et al.</a>
    develop a genome-scale model of all metabolic pathways and use this to
    analyze the efficacy of common cell treatments used to boost protein
    production. The model builds off the human genome-scale models RECON1 and
    RECON2, and has been extensively curated to ensure a high-quality resource
    for use in studying this model organism. The new model is in BiGG Models
    with the name
    <a href="/models/iCHOv1"><i>i</i>CHOv1</a>.
  </p>
  <a href="https://github.com/SBRG/bigg_models/releases/tag/v1.3.0">Github Release</a>
{% endblock %}
